File: OBI/Final2017/arranhaceu.py
from math import log2
import sys
sys.setrecursionlimit(2**20)
DEBUG = 0

def createSegmentTree(segmentTree, array, i, lo, hi):
    if (lo == hi):
        segmentTree[i] = array[lo]
        return(array[lo])
    elif (lo > hi):
        return(0)

    mid = (lo + hi) // 2
    segmentTree[i] = createSegmentTree(segmentTree, array, 2*i + 1, lo, mid) + createSegmentTree(segmentTree, array, 2*i + 2, mid + 1, hi)
    return(segmentTree[i])

def gs(segmentTree, i, lo, hi, qlo, qhi):
    if (lo >= qlo and hi <= qhi):
        return(segmentTree[i])
    elif (lo > qhi or hi < qlo):
        return(0)

    mid = (lo + hi) // 2
    return(gs(segmentTree, 2*i + 1, lo, mid, qlo, qhi) + gs(segmentTree, 2*i + 2, mid + 1, hi, qlo, qhi))

def getSum(segmentTree, size, lo, hi):
    if (lo < 0 or hi > size - 1 or lo > hi):
        return(-1)
    return(gs(segmentTree, 0, 0, size - 1, lo, hi))

def uv(segmentTree, diff, i, lo, hi, pos):
    if (lo > pos or hi < pos):
        return

    segmentTree[i] += diff
    if (lo != hi):
        mid = (lo + hi) // 2
        uv(segmentTree, diff, 2*i + 1, lo, mid, pos)
        uv(segmentTree, diff, 2*i + 2, mid + 1, hi, pos)

def updateValue(segmentTree, array, size, value, pos):
    if (pos < 0 or pos > size - 1):
        return
    diff = value - array[pos]
    array[pos] = value
    uv(segmentTree, diff, 0, 0, size - 1, pos)

floors, events = map(int, input().split())
building = list(map(int, input().split()))

segmentTree = [0] * (2 * 2**(int(log2(floors)) + 1) - 1)
createSegmentTree(segmentTree, building, 0, 0, floors - 1)

if (DEBUG):
    print("segmentTree", segmentTree)

for i in range(events):
    line = list(map(int, input().split()))
    if (line[0] == 0):
        floor, people = line[1:]
        updateValue(segmentTree, building, floors, people, floor - 1)
    else:
        floor = line[1]
        print(getSum(segmentTree, floors, 0, floor - 1))

File: Competitive Programming/OBI/Final2017/arranhaceu.cpp
#include <bits/stdc++.h>
#define lli long long int

int createSegmentTree(int segmentTree[], int array[], int i, int lo, int hi)
{
  if (lo == hi)
  {
    segmentTree[i] = array[lo];
    return(segmentTree[i]);
  }
  else if (lo > hi) return(0);

  int mid = (lo + hi) / 2;
  segmentTree[i] = createSegmentTree(segmentTree, array, 2*i + 1, lo, mid) + createSegmentTree(segmentTree, array, 2*i + 2, mid + 1, hi);
  return(segmentTree[i]);
}

int gs(int segmentTree[], int i, int lo, int hi, int qlo, int qhi)
{
  if (lo >= qlo && hi <= qhi)
    return(segmentTree[i]);
  else if (hi < qlo || lo > qhi)
    return(0);

  int mid = (lo + hi) / 2;
  return(gs(segmentTree, 2*i + 1, lo, mid, qlo, qhi) + gs(segmentTree, 2*i + 2, mid + 1, hi, qlo, qhi));
}

int getSum(int segmentTree[], int size, int lo, int hi)
{
  if (lo < 0 || hi > size - 1 || lo > hi)
    return(-1);

  return(gs(segmentTree, 0, 0, size - 1, lo, hi));
}

void uv(int segmentTree[], int i, int diff, int lo, int hi, int pos)
{
  if (lo > pos || hi < pos)
    return;

  segmentTree[i] += diff;
  if (lo != hi)
  {
    int mid = (lo + hi) / 2;
    uv(segmentTree, 2*i + 1, diff, lo, mid, pos);
    uv(segmentTree, 2*i + 2, diff, mid + 1, hi, pos);
  }
}

void updateValue(int segmentTree[], int array[], int size, int pos, int value)
{
  if (pos < 0 || pos > size - 1)
    return;

  int diff = value - array[pos];
  array[pos] = value;
  uv(segmentTree, 0, diff, 0, size - 1, pos);
}

int main()
{
  int floors, events; scanf("%d %d", &floors, &events);
  int building[floors], i;
  for (i = 0; i < floors; i ++)
    scanf("%d", &building[i]);

  int segmentTree[2 * (lli) pow(2, (lli) log2(floors) + 1) - 1];
  createSegmentTree(segmentTree, building, 0, 0, floors - 1);

  for (i = 0; i < events; i ++)
  {
    int kind; scanf("%d", &kind);
    if (kind)
    {
      int pos; scanf("%d", &pos);
      printf("%d\n", getSum(segmentTree, floors, 0, pos - 1));
    }
    else
    {
      int pos, people; scanf("%d %d", &pos, &people);
      updateValue(segmentTree, building, floors, pos - 1, people);
    }
  }

  return(0);
}